### PR TITLE
Add cleanup to folds

### PR DESCRIPTION
**Pull request by adithyaov**
- [x] Check regressions
- [x] Fix compile errors in tests and benchmarks
- [x] Fix formatting (<= 80 lines) (Only a few outliers)
- [ ] Add cleanup to `classifySessionsBy`

### CODE REVIEW

**Comment by adithyaov** [comment]
Regression report attached
[report.txt](https://github.com/composewell/streamly/files/5924078/report.txt)

File: src/Streamly/Internal/Data/Array/Prim/MutTypesInclude.hs [hunk 1]
```diff
@@ -378,6 +378,13 @@ lpackArraysChunksOf n (Fold step1 initial1 extract1) =
             FL.Partial rr -> extract1 rr
             FL.Done () -> return ()
 
+    clean (Tuple' Nothing r1) = clean1 r1
+    clean (Tuple' (Just buf) r1) = do
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Prim/MutTypesInclude.hs:382]: Just clean up directly here. Don't perform unnecessary actions.

File: src/Streamly/Internal/Data/Array/Stream/Mut/Foreign.hs [hunk 2]
```diff
@@ -156,6 +156,13 @@ lpackArraysChunksOf n (Fold step1 initial1 extract1) =
             FL.Partial rr -> extract1 rr
             FL.Done _ -> return ()
 
+    clean (Tuple' Nothing r1) = clean1 r1
+    clean (Tuple' (Just buf) r1) = do
+        r <- step1 r1 buf
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Stream/Mut/Foreign.hs:161]: Unnecessary action?

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 3]
```diff
@@ -395,6 +422,10 @@ mkFoldM = Fold
 mkFoldM_ :: Monad m => (b -> a -> m (Step b b)) -> m (Step b b) -> Fold m a b
 mkFoldM_ step initial = mkFoldM step initial return
 
+-- TODO
+-- 1. Rename mkFold-ish combinators
+-- 1. Add cleanup based smart constructors
+
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold/Types.hs:428]: Need suggestions on naming.

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 4]
```diff
@@ -434,10 +465,14 @@ toList = mkAccum (\f x -> f . (x :)) id ($ [])
 -- Instances
 ------------------------------------------------------------------------------
 
+-- XXX The reason I used "Monad m" is because it is used by "cleanup". I could
+-- have used applicative for nocleanup but I don't think it really matters.
+
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold/Types.hs:470]: Remove this line after review.

File: src/Streamly/Internal/Data/Parser/ParserD.hs [hunk 5]
```diff
@@ -166,6 +166,10 @@ import Prelude hiding
 import Streamly.Internal.Data.Parser.ParserD.Tee
 import Streamly.Internal.Data.Parser.ParserD.Types
 
+-- TODO Parsers will also have a cleanup action. The correct behaviour now is
+-- for the parsers to clean up the Fold resources while extracting. This is
+-- omitted now as parser will have a seperate cleanup later.
+
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Parser/ParserD.hs:172]: Remove after review.

Comment (adithyaov) [reply src/Streamly/Internal/Data/Parser/ParserD.hs:172]: We could clean up when extracting.

File: src/Streamly/Internal/Data/Stream/IsStream/Nesting.hs [hunk 6]
```diff
@@ -1751,6 +1751,7 @@ data SessionState t m k a b = SessionState
     , sessionOutputStream :: t (m :: Type -> Type) (k, b) -- ^ Completed sessions
     }
 
+-- XXX Need to check where to cleanup
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/IsStream/Nesting.hs:1754]: This is still incomplete. I Will get back to this or the reviewer can help me out if the info is known.

File: src/Streamly/Internal/Data/Stream/SVar.hs [hunk 7]
```diff
@@ -313,14 +313,16 @@ write svar winfo = Fold step initial extract
             void $ send svar (ChildYield x)
             return $ FL.Partial ()
 
-    extract () = liftIO $ sendStop svar winfo
+    extract () = return ()
+
+    cleanup () = liftIO $ sendStop svar winfo
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/SVar.hs:318]: See the effects

File: src/Streamly/Internal/Data/Stream/SVar.hs [hunk 8]
```diff
@@ -340,8 +342,10 @@ writeLimited svar winfo = Fold step initial extract
                 return $ FL.Done ()
     step False _ = return $ FL.Done ()
 
-    extract True = liftIO $ sendStop svar winfo
-    extract False = return ()
+    extract _ = return ()
+
+    cleanup True = liftIO $ sendStop svar winfo
+    cleanup False = return ()
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/SVar.hs:348]: See the effects of this.

File: src/Streamly/Internal/Network/Inet/TCP.hs [hunk 9]
```diff
@@ -352,21 +352,27 @@ writeChunks
     => (Word8, Word8, Word8, Word8)
     -> PortNumber
     -> Fold m (Array Word8) ()
-writeChunks addr port = Fold step initial extract
+writeChunks addr port = Fold step initial extract cleanup
```
Comment (adithyaov) [inline src/Streamly/Internal/Network/Inet/TCP.hs:355]: Cleanup internal fold in initial on exception?

File: src/Streamly/Internal/Data/Array/Prim/TypesInclude.hs [hunk 10]
```diff
@@ -615,14 +615,22 @@ lpackArraysChunksOf n (Fold step1 initial1 extract1) =
             FL.Partial rr -> extract1 rr
             FL.Done _ -> return ()
 
+    clean (Tuple3' Nothing' _ r1) = clean1 r1
+    clean (Tuple3' (Just' buf) boff r1) = do
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Prim/TypesInclude.hs:619]: unnecessary action?

**Comment by harendra-kumar** [comment]
Fixed by #2427

**Pull request closed without merging**